src/hooks/progress/use-update-tutorial-progress.ts

```diff
@@ -0,0 +1,186 @@
+import { useEffect } from 'react'
+import { useInView } from 'react-intersection-observer'
+import useAuthentication from 'hooks/use-authentication'
+// Tutorial progress utilities
+import { progressStatusToPercent } from 'lib/learn-client/api/progress/formatting'
+import {
+	useTutorialProgress,
+	useTutorialProgressMutations,
+	TutorialProgressMutationArgs,
+} from 'hooks/progress'
+// Types
+import {
+	TutorialIdCollectionId,
+	TutorialProgressStatus,
+} from 'lib/learn-client/types'
+
+type IntersectionRef = (node?: Element) => void
+
+interface UpdateProgressReturnInterface {
+	startRef: IntersectionRef
+	endRef: IntersectionRef
+}
+
+/**
+ * Given a tutorialId and collectionId,
+ * Return two refs in an object, { startRef, endRef },
+ * which can be placed on elements to track tutorial progress.
+ *
+ * On initial load, regardless of startRef or endRef visibility,
+ * if the tutorial does not yet have recorded progress,
+ * the tutorial will be marked as `visited` (0 percent).
+ *
+ * When the startRef element moves out of the viewport,
+ * the tutorial will be marked as `in_progress` (50 percent).
+ *
+ * When the endRef element moves into the viewport,
+ * the tutorial will be marked as `complete` (100 percent).
+ */
+export function useUpdateTutorialProgress({
+	tutorialId,
+	collectionId,
+}: TutorialIdCollectionId): UpdateProgressReturnInterface {
+	// We shouldn't try to update progress unless we're authenticated
+	const { isAuthenticated } = useAuthentication()
+	// We need to know if progress exists, to know whether to "create" or "update"
+	const { tutorialProgressStatus } = useTutorialProgress({
+		tutorialId,
+		collectionId,
+	})
+	// We'll use "create" or "update" progress using these `react-query` mutations
+	const tutorialProgressMutations = useTutorialProgressMutations()
+	// We use refs from `react-intersection-observer` to track progress.
+	const [startRef, startInView, startEntry] = useInView({
+		initialInView: true,
+	})
+	const [endRef, endInView, endEntry] = useInView()
+
+	/**
+	 * This effect runs often, but it seems necessary to ensure progress
+	 * gets tracked as things like the authenticated state changes.
+	 *
+	 * We bail early in many cases, and we avoid downgrading progress,
+	 * so this effect only results in meaningful mutations a small percent
+	 * of the time.
+	 */
+	useEffect(() => {
+		/**
+		 * On initial client-side navigation, this effect may be called
+		 * before startRefs and endRefs are updated. We can detect this
+		 * by comparing data-ref-id attributes we've added to the
+		 * elements startRef and endRef are attached to.
+		 */
+		const startRefId = startEntry?.target.getAttribute('data-ref-id')
+		const endRefId = endEntry?.target.getAttribute('data-ref-id')
+		const compareId = `${tutorialId}_${collectionId}`
+		const isStaticIdsMismatch =
+			startRefId !== compareId || endRefId !== compareId
```

Comment:
This is a key part of the effect that we'd been trying to debug on the ambient zoom yesterday. This seems to work exactly as expected now 👍

I've tried to add thorough yet brief comments on all the reasons we'd want to bail from this effect, as well as brief rationale for each effect dependency. Have formatted with early `return` so hopefully it's clearer than a single combined conditional.